### ISSUE TITLE
Allow plugins to access published plugin values via the plugin registrar/binding

### ISSUE THREAD

**Issue by fuzzybinary** [flutter/flutter#121527]
## Use case

When developing a plugin, or writing other native code, it is sometimes necessary to access the native implementation of another plugin. A good example of this is `webview_flutter`, which has `WebViewFlutterAndroidExternalApi` (Android) and `FWFWebViewFlutterWKWebViewExternalAPI` (iOS) classed for accessing the platform WebViews associated with a given handle. Both use either the `FlutterPluginRegistry` or the `FlutterEngine`, neither of which are easily accessible from a Plugin.

## Proposal

In my mind, the simplest solution that will have the smallest impact is to add properties onto the objects that are already passed to plugins during initialization.

iOS - Add a method to `FlutterPluginRegistrar` to allow upward access to the `FlutterPluginRegistry`

Android - Add a method to `FlutterPlugin.FlutterPluginBinding` to access the `FlutterEngine`

**Comment by stuartmorgan** [comment]
@bparrishMines Is there already a way to do this that I missed (in which case we just need a docs update)? Or is the current approach to require it to be wired up at the app level rather than the plugin level?

**Comment by bparrishMines** [comment]
We do need to track this feature because the way to access these classes in a plugin are not obvious and my solutions below are currently workarounds that are protected by tests. 

For iOS, the `FlutterAppDelegate` is an implementation of `FlutterPluginRegistry` and can be accessed through the application delegate

```objc
@import webview_flutter_wkwebview;

@interface MyPlugin ()
@property(weak) FlutterAppDelegate *appDelegate;
@end

@implementation MyPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
  FLTGoogleMobileAdsPlugin *instance = [[FLTGoogleMobileAdsPlugin alloc] init];
  // Adds plugin to listen for `application:didFinishLaunchingWithOptions:` below
  [registrar addApplicationDelegate:instance];
}

- (void)getWebView:(NSNumber *)webViewIdentifier {
  WKWebView *webView = (WKWebView *) [FWFWebViewFlutterWKWebViewExternalAPI
        webViewForIdentifier:webViewIdentifier.longValue       
          withPluginRegistry:self.appDelegate];
}


- (BOOL)application:(UIApplication *)application
   didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
  self.appDelegate = (FlutterAppDelegate *) application.delegate;
}
@end
```

For Android you can access `FlutterEngine` through `FlutterPluginBinding`. However I would note that `FlutterPluginBinding.getFlutterEngine()` is currently deprecated. We don't currently have a plan to remove it, but we may at some point. An alternative solution to access other plugins will be provided before that happens though. 

```java
void getWebView(long webViewIdentifier) {
 final FlutterPluginBinding binding = /*get saved binding*/;
 final WebView webView = WebViewFlutterAndroidExternalApi.getWebView(
     binding.getFlutterEngine(), webViewIdentifier);
}
```

**Comment by fuzzybinary** [comment]
These are exactly the workarounds I'm using at the moment. On the iOS side, the issue that made me raise this is trying to use the plugin registry in an "add to app" scenario, where we can't assume The AppDelegate is a FlutterAppDeleage.

**Comment by stuartmorgan** [comment]
I don't think we probably want to go as far as exposing the whole registry or engine to every plugin, since the registrar is intended to be a limited interface. Making breaking changes to an interface becomes *much* harder once plugins are using it.

Instead we could plumb through the specific call to get the published value from another plugin.

**Comment by msbit** [comment]
We're faced with a similar scenario to @fuzzybinary, in that we need a reference to the `FlutterPluginRegistry` from an "add to app" scenario, in our case using the `google_mobile_ads` package. I certainly may be missing something, but I don't see a way to proceed with that as is stands. I would be very happy to be proven wrong, of course :D

**Comment by abhinavsingh** [comment]
We already do this on iOS using `valuePublished` utility method.   But on Android, `FlutterPluginRegistry` has now been marked as deprecated.   It does contain a method named `valuePublishedByPlugin` which is exactly what we need. 

1) Is there a recommended workaround for Android?
2) By any chance, is Flutter planning to deprecate `valuePublished` we have on iOS platform?

**Comment by abhinavsingh** [comment]
> We already do this on iOS using [`valuePublished`](https://stackoverflow.com/questions/69064383/handling-userintent-in-flutter-by-calling-eventsink-from-appdelegate) utility method. But on Android, `FlutterPluginRegistry` has now been marked as deprecated. It does contain a method named `valuePublishedByPlugin` which is exactly what we need.
> 
> 1. Is there a recommended workaround for Android?
> 2. By any chance, is Flutter planning to deprecate `valuePublished` we have on iOS platform?

Whats unclear to me is:

1) How do I get access to `FlutterPluginRegistry` instance on Android
2) Is there a recommended workaround to access another plugin instance for Android

In my case, I want to delegate processing of an intent from `MainActivity` to a plugin, which already implements handling for this intent.

**Comment by stuartmorgan-g** [comment]
> I'm wondering if this can be revisited

What specifically needs to be revisited? The discussion here still seems to be valid. In particular:

> It would be nice if FlutterPlugins were just given a way to access the registry they're a part of.

All of the use cases given here would be addressed by [my proposal above](https://github.com/flutter/flutter/issues/121527#issuecomment-1446821540), so I'm not sure why we need to re-litigate exposing the entire registry.

**Comment by fuzzybinary** [comment]
Sorry, I'm dealing with trying to find a workaround to this issue as it exists, so I'm a bit frustrated. My apologies 😓.

I guess I'm more asking if we can get this added, more than revisiting the proposal. A method to access the published value from another Plugin is really all I want. Where would that be added though? The only thing Plugins can really see is the Registrar. Is that were it should live?

**Comment by fuzzybinary** [comment]
Also, note that [this method](https://github.com/flutter/packages/blob/3b856a2ab061792e3e8b206efc41649c69078812/packages/webview_flutter/webview_flutter_wkwebview/darwin/webview_flutter_wkwebview/Sources/webview_flutter_wkwebview/WebViewFlutterWKWebViewExternalAPI.swift#L28) is the main feature of `WebviewFlutterPlugin` that I'm trying to access, which takes a FlutterPluginRegistry, which is why I guess I'm caught up in trying to access the Registry rather than the value published by the Plugin.
